### ISSUE TITLE
Temporal Dead Zone ReferenceError rule

### ISSUE THREAD

**Issue by asapach** [eslint/eslint#9677]
* **ESLint Version:** 4.12.1

```json
{
    "parserOptions": {
        "ecmaVersion": 6
    },
    "rules": {
        "no-use-before-define": "error"
    }
}
```

Example taken from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let#Temporal_Dead_Zone_and_errors_with_let:

```js
function do_something() {
  console.log(bar); // undefined
  console.log(foo); // ReferenceError
  var bar = 1;
  let foo = 2;
}
```

```bash
  2:17  error  'bar' was used before it was defined  no-use-before-define
  3:17  error  'foo' was used before it was defined  no-use-before-define
```

**What did you expect to happen?**
While both of these errors are valid, they are not of equal severity: usage of `bar` is a *potential* coding mistake, whereas `foo` is ***guaranteed*** to cause a `ReferenceError` when `do_something()` is executed:
* Chrome/Node/V8: `ReferenceError: foo is not defined`
* Firefox: ``ReferenceError: can't access lexical declaration `foo' before initialization``
* IE/Edge: `SCRIPT5113: Use before declaration`

There should be a way to distinguish between the two. [`no-use-before-define`](https://eslint.org/docs/rules/no-use-before-define) is the only way to catch this, but it doesn't allow to skip hoisted variables and only report TDZ errors.

**What actually happened?**
#7111 introduced [`variables` option](https://eslint.org/docs/rules/no-use-before-define#options), but it doesn't allow to distinguish between `var` and `let`/`const`.

There should either be a new rule that detects TDZ errors or `no-use-before-define` should be extended to allow to distinguish between hoisted and block-scoped vars, like it does for functions (which are hoisted) and classes (which are not). I would also suggest reporting TDZ errors in `eslint:recommended` by default.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
